impl/src/main/java/org/jboss/weld/util/Beans.java

```diff
@@ -640,4 +644,44 @@ public static boolean hasBuiltinScope(Bean<?> bean) {
         manager.getServices().get(SlimAnnotatedTypeStore.class).put(implementationClass.slim());
         return implementationClass;
     }
+
+    static ImmutableSet.Builder<Type> omitIllegalBeanTypes(Set<Type> types, EnhancedAnnotated<?, ?> annotated) {
+        ImmutableSet.Builder<Type> builder = ImmutableSet.builder();
```

Comment:
This might be a premature optimization but what about this: Since the case of an illegal bean type is very rare (the input set is the same as the returned one most of the time) what about by default only iterating over the types and checking if any of them is invalid. If an invalid type is found only then start building the new set.